File: src/main/java/com/github/sblundy/elvish/psi/ElvishChunkBase.java
package com.github.sblundy.elvish.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class ElvishChunkBase extends ASTWrapperPsiElement implements ElvishChunk, ElvishVariableScope {
    public ElvishChunkBase(@NotNull ASTNode node) {
        super(node);
    }

    @NotNull
    @Override
    public Collection<ElvishVariableDeclaration> findVariables(@NotNull String name, @NotNull List<String> ns) {
        List<ElvishVariableDeclaration> found = new ArrayList<>();
        for (ElvishAssignment assignment : getAssignmentList()) {
            for (ElvishVariableDeclaration declaration : assignment.getVariableList()) {
                if (declaration.nameMatches(name, ns)) {
                    found.add(declaration);
                }
            }
        }

        found.addAll(ElvishPsiUtils.INSTANCE.findVariableInParentScope(name, ns, this));
        return found;
    }
}

File: src/main/kotlin/com/github/sblundy/elvish/psi/ElvishChunkBase.kt
package com.github.sblundy.elvish.psi

import com.github.sblundy.elvish.psi.ElvishPsiUtils.findFnCommandInParentScope
import com.github.sblundy.elvish.psi.ElvishPsiUtils.findVariableInParentScope
import com.github.sblundy.elvish.psi.ElvishPsiUtils.processFnCommandInParentScope
import com.github.sblundy.elvish.psi.ElvishPsiUtils.processVariablesInParentScope
import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode

abstract class ElvishChunkBase(node: ASTNode) : ASTWrapperPsiElement(node), ElvishChunk, ElvishDeclarationScope {
    override fun findVariables(ref: ReferenceWithNamespacePsiElement): Collection<ElvishVariableDeclaration> {
        return assignmentList.flatMap { it.variableList }
            .filter { it.nameMatches(ref) } +
                findVariableInParentScope(ref, this)
    }

    override fun processVariables(processor: ElvishVariableScope.VariableProcessor) {
        for (v in assignmentList.flatMap { it.variableList }) {
            if (!processor.process(v)) {
                return
            }
        }

        processVariablesInParentScope(processor, this)
    }

    override fun findFnCommands(ref: ReferenceWithNamespacePsiElement): Collection<ElvishFunctionDeclaration> {
        return fnCommandList.filter { it.nameMatches(ref) } + findFnCommandInParentScope(ref, this)
    }

    override fun processFnCommands(processor: ElvishFunctionScope.FnCommandProcessor) {
        for (v in fnCommandList) {
            if (!processor.process(v)) {
                return
            }
        }

        processFnCommandInParentScope(processor, this)
    }
}